File: frameworks/rust/poem/Implementation/routes/template.rs
use std::sync::Arc;

use poem::endpoint::make_sync;
use poem::error::InternalServerError;
use poem::web::Html;
use poem::{Result, Route, get};
use tera::{Context, Tera};

use crate::Payloads;
use crate::payloads::Payload;

// rb:wiring template.*
/// Implementation/templates/items.html, compiled into the binary and parsed by Tera once, when the
/// application is built. poem has no view layer, and its own templating example renders with Tera.
fn templates() -> Tera {
    let mut tera = Tera::new();
    tera.add_raw_template("items.html", include_str!("../templates/items.html")).expect("items.html is a Tera template");
    tera
}

/// The page for one payload. Tera renders from a context, so the payload is serialised into one on
/// every request.
fn page(tera: &Tera, body: &Payload) -> Result<Html<String>> {
    let mut context = Context::new();
    context.insert("body", body);
    tera.render("items.html", &context).map(Html).map_err(InternalServerError)
}
// rb:end

/// template: the payloads rendered as a page.
pub fn add(route: Route, p: &'static Payloads) -> Route {
    let small = Arc::new(templates());
    let medium = small.clone();
    route
        .at("/template/small", get(make_sync(move |_| page(&small, &p.small))))
        .at("/template/medium", get(make_sync(move |_| page(&medium, &p.medium))))
}

File: frameworks/go/huma/Implementation/template.go
package implementation

import (
	"bytes"
	"context"
	"embed"
	"html/template"

	"github.com/danielgtaylor/huma/v2"
)

// rb:wiring template.*

//go:embed views/items.tmpl
var views embed.FS

// templateRoutes render the payload with html/template into a buffer, and answer it as a []byte
// body under text/html, as Huma's HTML response page does. Huma has no view layer. The template
// is compiled into the binary and parsed once.
func templateRoutes(api huma.API, p *Payloads) {
	// rb:wiring template.*
	page := template.Must(template.ParseFS(views, "views/items.tmpl"))
	rendered := func(payload *Payload) func(context.Context, *struct{}) (*TextOutput, error) {
		return func(ctx context.Context, _ *struct{}) (*TextOutput, error) {
			var html bytes.Buffer
			if err := page.Execute(&html, payload); err != nil {
				return nil, huma.Error500InternalServerError("the page does not render", err)
			}
			return &TextOutput{ContentType: "text/html; charset=utf-8", Body: html.Bytes()}, nil
		}
	}
	// rb:end

	huma.Get(api, "/template/small", rendered(&p.Small))

	huma.Get(api, "/template/medium", rendered(&p.Medium))
}
